Introduce an ApplicationInstance-specific registry to Boot Cleanup WIP

dgeb opened the pull request:
The ApplicationInstance-specific registry uses its corresponding Application's registry as a failover to resolve registrations and injections.

Soon, this registry will be abstracted behind proxies such as `register` and `inject` on the instance itself.

cc @wycats @tomdale

tomdale commented on packages/container/lib/registry.js:
```diff
@@ -18,23 +18,34 @@ var VALID_FULL_NAME_REGEXP = /^[^:]+.+:[^:]+$/;
  @class Registry
 */
 function Registry(options) {
+  this.failover = options && options.failover ? options.failover : null;
```
This is a total bikeshed, but I'm not _totally_ in love with the term "failover." To me, a failover implies latch-like behavior: once the "fail" has occurred, you switch everything over to the failover. For example, if the power goes out, you have a "failover" generator that everything switches to.

In this case, it's not really a failover per se… it just looks in its local one first, then expands its search to the parent registry if it's available. Perhaps this terminology is controversial, but I think I'd almost prefer `parentRegistry` to `failover`.

rwjblue replied:
fallback?

tomdale replied:
Fallback seems good.

dgeb replied:
ok, `fallback` it is - thanks @rwjblue!

dgeb committed:
Create a registry per ApplicationInstance

The instance's registry sets its fallback to the corresponding
application's registry.

dgeb committed:
Introduce fallback registry.

A registry's fallback registry resolves registrations and injections
when no matches can be found.

dgeb commented:
@tomdale @rwjblue ok, I've fixed up my commits to use `fallback` instead of `failover`

dgeb committed:
Fix module name

dgeb committed:
Move validation of `fullName` parsing to `_parseName`.

Ensure that `name` and `type` are valid after initial parsing and before
any post-parsing mutations might be performed.

This fixes a problem in which mutations are performed to the cached
object (e.g. "route:basic" -> "route:") and then the validation of the
mutated object fail when it's accessed a second time.

It's also more efficient, since validation only occurs once.

tomdale commented:
Bless you child. :pray:

dgeb commented:
:beers:
